### DISCUSSION TITLE
Visual Studio Code reports NimBLE-Arduino library problem while compiling 13.1.0.4 version

### DISCUSSION THREAD

**Discussion by jpier123** [arendst/Tasmota#19792] [Q&A] [answered]
### PROBLEM DESCRIPTION
It was disovered  NimBLE-Arduino library problem while compiling 13.1.0.4 version using Visual Studio Code. There was no such a problem when used 13.1.0.1 dev build.

platformio.ini
##################################################
[env:tasmota32_base]
framework = ${common.framework}
platform = ${core32.platform}
platform_packages = ${core32.platform_packages}
board_build.filesystem = ${common.board_build.filesystem}
custom_unpack_dir = ${common.custom_unpack_dir}
board = esp32
monitor_speed = 115200
upload_resetmethod = ${common.upload_resetmethod}
extra_scripts = ${esp32_defaults.extra_scripts}
build_unflags = ${core32.build_unflags}
build_flags = ${core32.build_flags}
lib_ldf_mode = ${common.lib_ldf_mode}
lib_compat_mode = ${common.lib_compat_mode}
lib_extra_dirs = ${common.lib_extra_dirs}
	lib/libesp32
	lib/libesp32_lvgl
	lib/libesp32_audio
lib_ignore = 
	HTTPUpdateServer
	ESP RainMaker
	WiFiProv
	USB
	SPIFFS
	ESP32 Azure IoT Arduino
	ESP32 Async UDP
	ESP32 BLE Arduino
	SimpleBLE
	NetBIOS
	ESP32
	Preferences
	BluetoothSerial
	ArduinoOTA
custom_files_upload = no_files
lib_deps = h2zero/NimBLE-Arduino@^1.4.1
##################################################

lib_deps = h2zero/NimBLE-Arduino@^1.4.1
Used PlatformIO CLI commnad: platformio run -e tasmota32

PlatformIO CLI output
##################################################

**_C:/Users/Administrator/Documents/PlatformIO/Projects/Tasmota-dev v13.1.0.4/tasmota/tasmota_xsns_sensor/xsns_62_esp32_mi.ino:122:52: error: expected class-name before '{' token
 class MI32AdvCallbacks: public NimBLEScanCallbacks {_**
                                                                                     ^
In file included from .pio/libdeps/tasmota32/NimBLE-Arduino/src/nimble/esp_port/port/include/esp_nimble_cfg.h:4,
                 from .pio/libdeps/tasmota32/NimBLE-Arduino/src/nimble/porting/nimble/include/syscfg/syscfg.h:9,
                 from .pio/libdeps/tasmota32/NimBLE-Arduino/src/nimble/nimble/host/mesh/src/shell.c:12:
.pio/libdeps/tasmota32/NimBLE-Arduino/src/nimconfig.h:158: warning: "CONFIG_BT_NIMBLE_ROLE_OBSERVER" redefined
 #define CONFIG_BT_NIMBLE_ROLE_OBSERVER

C:/Users/Administrator/Documents/PlatformIO/Projects/Tasmota-dev v13.1.0.4/tasmota/tasmota_xsns_sensor/xsns_62_esp32_mi.ino: In function 'void MI32ScanTask(void*)':
**_C:/Users/Administrator/Documents/PlatformIO/Projects/Tasmota-dev v13.1.0.4/tasmota/tasmota_xsns_sensor/xsns_62_esp32_mi.ino:1205:13: error: 'class NimBLEScan' has no member named 'setScanCallbacks'_**
     if (MI32.mode.connected == 1) return;
                 ^~~~~~~~~~~~~~~~
In file included from <command-line>:
C:/Users/Administrator/.platformio/packages/framework-arduinoespressif32/tools/sdk/esp32/dio_qspi/include/sdkconfig.h:127: note: this is the location of the previous definition
 #define CONFIG_BT_NIMBLE_ROLE_OBSERVER 1

*** [.pio\build\tasmota32\src\tasmota.ino.cpp.o] Error 1
##################################################

### REQUESTED INFORMATION
_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [x] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [x] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [x] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [x] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [x] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [x] Device used (e.g., Sonoff Basic): ESP32-S2 Mini
- [x] Tasmota binary firmware version number used: 13.1.0.4
  - [ ] Pre-compiled
  - [x] Self-compiled
- [ ] Flashing tools used: _____
- [ ] Provide the output of command: `Backlog Template; Module; GPIO 255`:
```lua
  Configuration output here:


```
- [ ] If using rules, provide the output of this command: `Backlog Rule1; Rule2; Rule3`:
```lua
  Rules output here:


```
- [ ] Provide the output of this command: `Status 0`:
```lua
  STATUS 0 output here:


```
- [ ] Set `weblog` to 4 and then, when you experience your issue, provide the output of the Console log:
```lua
  Console output here:


```

### TO REPRODUCE
_Steps to reproduce the behavior:_

### EXPECTED BEHAVIOUR
_A clear and concise description of what you expected to happen._

### SCREENSHOTS
_If applicable, add screenshots to help explain your problem._

### ADDITIONAL CONTEXT
_Add any other context about the problem here._

**(Please, remember to close the issue when the problem has been addressed)**

**Comment by Staars** [comment, marked as answer]
You added a dependency on your own:
lib_deps = h2zero/NimBLE-Arduino@^1.4.1

Such changes are undefined behaviour in the scope of Tasmotas build system.

**Comment by Jason2866** [comment]
h2zero NimBLE is included in the Tasmota libs. Tasmota don't use ANY external libs. Most libs are modified for use with Tasmota. So it is for NimBLE too. Tasmota uses the IDF variant of this lib. Reason it is more actual and does already support the C2 and C6.

**Comment by jpier123** [comment]
Hi Jason,

I'm almost 100% sure that both 13.1.0.4 and 13.2.0 build versions runing with the same bug when USE_MI_ESP32 was in use.

I have fresh 13.2.0 build with no lh2zero/NimBLE-Arduino@^1.4.1 dependency.. 

Please note that I use USE_MI_ESP32  BLE-bridge,please see below relevant my_user_config.h part:

#define USE_MI_ESP32                             // Add support for ESP32 as a BLE-bridge (+9k2 mem, +292k flash)
//#define USE_BLE_ESP32                            // Add support for ESP32 as a BLE-bridge (+9k2? mem, +292k? flash)
#define BLE_ESP32_ENABLE false                   //-- [SetOption115] Default value for SetOption115

Please note that error does not appear when USE_MI_ESP32  BLE-bridge was not definied,please see below relevant my_user_config.h part:

//#define USE_MI_ESP32                             // Add support for ESP32 as a BLE-bridge (+9k2 mem, +292k flash)
#define USE_BLE_ESP32                            // Add support for ESP32 as a BLE-bridge (+9k2? mem, +292k? flash)
#define BLE_ESP32_ENABLE false                   //-- [SetOption115] Default value for SetOption115

##############################################################

Compiling .pio\build\tasmota32\lib0cf\lvgl\draw\lv_draw.c.o
Compiling .pio\build\tasmota32\lib0cf\lvgl\draw\lv_draw_arc.c.o
Compiling .pio\build\tasmota32\lib0cf\lvgl\draw\lv_draw_img.c.o
Compiling .pio\build\tasmota32\lib0cf\lvgl\draw\lv_draw_label.c.o
C:/Users/Administrator/Documents/PlatformIO/Projects/Tasmota v13.2.0/tasmota/tasmota_xsns_sensor/xsns_62_esp32_mi.ino:111:52: error: expected class-name before '{' token
 class MI32AdvCallbacks: public NimBLEScanCallbacks {
                                                    ^
C:/Users/Administrator/Documents/PlatformIO/Projects/Tasmota v13.2.0/tasmota/tasmota_xsns_sensor/xsns_62_esp32_mi.ino: In function 'void MI32ScanTask(void*)':
C:/Users/Administrator/Documents/PlatformIO/Projects/Tasmota v13.2.0/tasmota/tasmota_xsns_sensor/xsns_62_esp32_mi.ino:1050:13: error: 'class NimBLEScan' has no member named 'setScanCallbacks'
 void MI32ScanTask(void *pvParameters){
             ^~~~~~~~~~~~~~~~
Compiling .pio\build\tasmota32\lib0cf\lvgl\draw\lv_draw_layer.c.o
Compiling .pio\build\tasmota32\lib0cf\lvgl\draw\lv_draw_line.c.o
Compiling .pio\build\tasmota32\lib0cf\lvgl\draw\lv_draw_mask.c.o
Compiling .pio\build\tasmota32\lib0cf\lvgl\draw\lv_draw_rect.c.o
Compiling .pio\build\tasmota32\lib0cf\lvgl\draw\lv_draw_transform.c.o
Compiling .pio\build\tasmota32\lib0cf\lvgl\draw\lv_draw_triangle.c.o
Compiling .pio\build\tasmota32\lib0cf\lvgl\draw\lv_img_buf.c.o
At global scope:
cc1plus.exe: warning: unrecognized command line option '-Wno-volatile'
*** [.pio\build\tasmota32\src\tasmota.ino.cpp.o] Error 1
========================================================================================= [FAILED] Took 124.38 seconds =========================================================================================

Environment    Status    Duration
-------------  --------  ------------
tasmota32      FAILED    00:02:04.376
===================================================================================== 1 failed, 0 succeeded in 00:02:04.376 =====================================================================================PS C:\Users\Administrator\Documents\PlatformIO\Projects\Tasmota v13.2.0> 

##############################################################

**Comment by Jason2866** [comment]
All this env do compile successfully: https://github.com/Jason2866/Tasmota-build/blob/special_builder/platformio_tasmota_cenv.ini#L250-L353

**Comment by barbudor** [comment]
@jpier123 I always compile USE_MI_ESP32 + USE_BLE_ESP32 as the sample pointed by Jason
You need BLE for the MI sensors

**Reply by jpier123** [reply]
I understood, Thank you both.
